src/madsci_client/madsci/client/data_client.py

```diff
@@ -34,41 +34,149 @@ def __init__(
+                secret_key=self.object_storage_config.secret_key,
+                secure=self.object_storage_config.secure,
+            )
+
+            # Ensure the default bucket exists
+            if not self._minio_client.bucket_exists(
+                self.object_storage_config.default_bucket
+            ):
+                self._minio_client.make_bucket(
+                    self.object_storage_config.default_bucket
+                )
+
+        except Exception as e:
+            warnings.warn(
+                f"Failed to initialize object storage client: {e!s}",
+                UserWarning,
+                stacklevel=2,
+            )
+            self._minio_client = None
+
+    def _get_datapoint_metadata(self, datapoint_id: Union[str, ULID]) -> DataPoint:
+        """Internal method to get datapoint metadata, either from local storage or server."""
         if self.url is None:
-            return self._local_datapoints[datapoint_id]
+            if datapoint_id in self._local_datapoints:
+                return self._local_datapoints[datapoint_id]
+            raise ValueError(f"Datapoint {datapoint_id} not found in local storage")
+
         response = requests.get(f"{self.url}datapoint/{datapoint_id}", timeout=10)
         response.raise_for_status()
         return DataPoint.discriminate(response.json())
 
+    def get_datapoint(self, datapoint_id: Union[str, ULID]) -> DataPoint:
+        """Get a datapoint by ID."""
+        return self._get_datapoint_metadata(datapoint_id)
+
     def get_datapoint_value(self, datapoint_id: Union[str, ULID]) -> Any:
-        """Get an datapoint value by ID. If the datapoint is JSON, returns the JSON data.
+        """Get a datapoint value by ID. If the datapoint is JSON, returns the JSON data.
         Otherwise, returns the raw data as bytes."""
-        if self.url is None:
-            if hasattr(self._local_datapoints[datapoint_id], "value"):
-                return self._local_datapoints[datapoint_id].value
-            if hasattr(self._local_datapoints[datapoint_id], "path"):
-                with (
-                    Path(self._local_datapoints[datapoint_id].path)
-                    .resolve()
-                    .expanduser()
-                    .open("rb") as f
-                ):
-                    return f.read()
-        response = requests.get(f"{self.url}datapoint/{datapoint_id}/value", timeout=10)
-        response.raise_for_status()
-        try:
-            return response.json()
-        except JSONDecodeError:
-            return response.content
+        # First get the datapoint metadata
+        datapoint = self.get_datapoint(datapoint_id)
+        # Handle based on datapoint type (regardless of URL configuration)
+        if self._minio_client is not None:
+            # Use MinIO client if configured
+            if (
+                hasattr(datapoint, "data_type")
+                and str(datapoint.data_type.value) == "object_storage"
```

Comment:
These kinds of checks should probably use the DataPointTypeEnum instead of a raw string comparison, or an isinstance check since we should have done a `.discriminate` call when we got the datapoint